crates/ra_hir_ty/src/lib.rs

```diff
@@ -662,13 +662,27 @@ impl TypeWalk for GenericPredicate {
 
 /// Basically a claim (currently not validated / checked) that the contained
 /// type / trait ref contains no inference variables; any inference variables it
-/// contained have been replaced by bound variables, and `num_vars` tells us how
-/// many there are. This is used to erase irrelevant differences between types
-/// before using them in queries.
+/// contained have been replaced by bound variables, and `kinds` tells us how
+/// many there are and whether they were normal or float/int variables. This is
+/// used to erase irrelevant differences between types before using them in
+/// queries.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct Canonical<T> {
     pub value: T,
-    pub num_vars: usize,
+    pub kinds: Arc<[TyKind]>,
```

Comment:
I wonder if it makes sense to micro-optimize here: `kinds: Either<u64, ThinArc<[TyKind]>>`. It maybe not worth doing this right now, but we can future proof, by introducing `struct Kinds(Arc<[TyKind]>)` with some boilerplate methods, just so we can easily tweak it later.

Reply:
This will get replaced by the Chalk version anyway (where we then can micro-optimize the interned representation), so I don't know if it's worth doing now.